### PR TITLE
Feature/313 get user projects

### PR DESCRIPTION
**Pull request by waltersajtos**
## Description

For the project profile page we are going to need an endpoint to fetch the projects that belong to the user

## Type of change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I did not update API Controllers, if I did, I added/changed Postman tests
-   [x] I have performed a self-review of my own code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [ ] I have made corresponding changes to the documentation
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes
-   [ ] I updated the changelog with an end-user readable description
-   [x] I assigned this pull request to the correct project board to update the sprint board

## Steps to Test or Reproduce

Outline the steps to test or reproduce the PR here.
These steps will be used during release testing.

1. Create a project on a user
2. Make a request to {dexAPI}/user/projects
3. The API should return the user's projects

## Link to issue
#313 

Closes: #313

### CODE REVIEW

File: API/Controllers/UserController.cs [hunk 1]
```diff
@@ -148,15 +153,45 @@ public async Task<IActionResult> GetUser(int userId)
...
+            /// <summary>
+            /// This method is responsible for creating the account.
+            /// </summary>
+            /// <param name="accountResource">The account resource which is used for creating the account.</param>
+            /// <returns>This method returns the created user as user resource result.</returns>
+            /// <response code="200">This endpoint returns the created user.</response>
+            /// <response code="400">The 400 Bad Request status code is return when the institution id is invalid
+            /// or when saving the user to the database failed.</response>
+            /// <response code="404">The institution with the specified institution id could not be found.</response>
+            [HttpPost]
```
Comment (Brend-Smits) [inline API/Controllers/UserController.cs:194]: Formatting is off?

File: API/Controllers/UserController.cs [hunk 2]
```diff
@@ -148,15 +153,45 @@ public async Task<IActionResult> GetUser(int userId)
...
+        /// <response code="200">This endpoint returns the current users projects.</response>
+        /// <response code="404">The 404 Not found status code is returned when the user could not be found.</response>
+        [HttpGet("projects")]
+        [Authorize]
+        [ProducesResponseType(typeof(UserResourceResult), (int) HttpStatusCode.OK)]
+        [ProducesResponseType(typeof(ProblemDetails), (int) HttpStatusCode.NotFound)]
+        public async Task<IActionResult> GetUserProjects()
+        {
+            string identityId = HttpContext.User.GetIdentityId(HttpContext);
+            User user = await userService.GetUserByIdentityIdAsync(identityId);
```
Comment (Brend-Smits) [inline API/Controllers/UserController.cs:168]: I think you can just use the helper method GetContextUser

File: Repositories/ProjectRepository.cs [hunk 3]
```diff
@@ -321,5 +323,20 @@ private IQueryable<Project> GetProjectQueryable(string query)
...
+
+        /// <summary>
+        /// Get the user projects.
+        /// </summary>
+        /// <param name="userId">The id of the user whoms projects need to be retrieved</param>
+        /// <returns>A enumerable of the users projects</returns>
+        public async Task<IEnumerable<Project>> FindUserProjectsWithUserAndCollaboratorsAsync(int userId)
+        {
+            IEnumerable<Project> projects = await GetDbSet<Project>()
+                   .Include(p => p.Collaborators)
```
Comment (Brend-Smits) [inline Repositories/ProjectRepository.cs:335]: Refer to my PR #335 please

File: Repositories/ProjectRepository.cs [hunk 4]
```diff
@@ -321,5 +323,20 @@ private IQueryable<Project> GetProjectQueryable(string query)
                             p.Id.ToString().Equals(query) ||
                             p.User.Name.Contains(query));
         }
+
+        /// <summary>
+        /// Get the user projects.
+        /// </summary>
+        /// <param name="userId">The id of the user whoms projects need to be retrieved</param>
+        /// <returns>A enumerable of the users projects</returns>
+        public async Task<IEnumerable<Project>> FindUserProjectsWithUserAndCollaboratorsAsync(int userId)
```
Comment (Brend-Smits) [inline Repositories/ProjectRepository.cs:332]: FindUserProjectsAsync is sufficient I think

File: Services/Services/ProjectService.cs [hunk 5]
```diff
@@ -144,6 +150,10 @@ public Task<Project> FindWithUserAndCollaboratorsAsync(int id)
             return Repository.FindWithUserAndCollaboratorsAsync(id);
         }
 
+        public Task<IEnumerable<Project>> GetUserProjects(int userId)
+        {
+            return Repository.FindUserProjectsWithUserAndCollaboratorsAsync(userId);
```
Comment (Brend-Smits) [inline Services/Services/ProjectService.cs:155]: FindUserProjectsAsync is sufficient

**Review by Brend-Smits — CHANGES_REQUESTED** [review]

File: API/Controllers/UserController.cs [hunk 6]
```diff
@@ -55,19 +57,22 @@ public class UserController : ControllerBase
         /// <param name="institutionService">The institution service which is used to communicate with the logic layer.</param>
         /// <param name="authorizationHelper">The authorization helper which is used to communicate with the authorization helper class.</param>
         /// <param name="userUserService">The user user service is responsible for users that are following users.</param>
+        /// <param name="projectService">The project service which is responsible for retreiving the users projects</param>
```
Comment (niraymak) [inline API/Controllers/UserController.cs:60]: Typo: retreiving -> retrieving

**Review by niraymak — CHANGES_REQUESTED** [review]

**Comment by waltersajtos** [comment]
I still need to fix the merge conflicts for the tests, I will do this once #77 and #354 are merged so we don't need to fix the merge conflicts there

File: Repositories/ProjectRepository.cs [hunk 7]
```diff
@@ -281,6 +287,29 @@ public override void Update(Project entity)
...
+                   .Include(p => p.ProjectIcon)
+                   .Where(p => p.UserId == userId)
+                   .ToListAsync();
+
+            foreach(Project project in projects)
+            {
+                project.Collaborators = await GetDbSet<Collaborator>()
+                                              .Where(p => p.ProjectId == project.Id)
+                                              .ToListAsync();
+            }
```
Comment (niraymak) [inline Repositories/ProjectRepository.cs:308]: Collaborators are already included on line 298, right? I think these lines can be removed.

**Review by niraymak — CHANGES_REQUESTED** [review]
Great work! 

I found some redundant code which slows down our application. You can remove those the lines. I already tested it and it works.

File: Repositories/ProjectRepository.cs (hunk 7 above)
Comment (niraymak) [reply Repositories/ProjectRepository.cs:308]: > Great work!
> 
> I found some redundant code which slows down our application. You can remove those the lines. I already tested it and it works.

And don't forget to change the changelog

File: Postman/dex.postman_collection.json [hunk 8]
```diff
@@ -8597,7 +8597,7 @@
 									"response": []
 								},
 								{
-									"name": "Project-DeleteProject-Other-Registered",
+									"name": "Project-DeleteProject-Other-Registered Copy",
```
Comment (Ruby77) [inline Postman/dex.postman_collection.json:8600]: Replace copy by an actual name

**Review by Ruby77 — CHANGES_REQUESTED** [review]
Looks good! I'll approve after you resolved Niray's feedback :smile:

File: CHANGELOG.md [hunk 9]
```diff
@@ -10,8 +10,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 
 ### Added
+
 - Automatically link users to their institution - [#295](https://github.com/DigitalExcellence/dex-backend/issues/295)
 - Collaborators are now included on the project overview page - [#317](https://github.com/DigitalExcellence/dex-backend/issues/317)
+- Added functionality to retrieve a users project - [#313](https://github.com/DigitalExcellence/dex-backend/issues/313)
```
Comment (Ruby77) [inline CHANGELOG.md]: Final request change :smile: . I will approve after this change!

```suggestion
- Added functionality to retrieve a user's project - [#313](https://github.com/DigitalExcellence/dex-backend/issues/313)
```

File: CHANGELOG.md [hunk 10]
```diff
@@ -40,6 +40,7 @@ Graduation system for users whose FHICT account becomes unavailable. [#287](http
 ### Added
 
 - Added ability to make projects private to a specific institution - [#310](https://github.com/DigitalExcellence/dex-backend/issues/310)
+- Added ability to retreive all user projects - [#313](https://github.com/DigitalExcellence/dex-backend/issues/313)
```
Comment (Ruby77) [inline CHANGELOG.md:43]:
```suggestion
- Added ability to retrieve all the user's projects - [#313](https://github.com/DigitalExcellence/dex-backend/issues/313)
```

**Review by Ruby77 — CHANGES_REQUESTED** [review]

**Review by Ruby77 — APPROVED** [review]
Good work. Everything works as expected!

**Review by niraymak — APPROVED** [review]
💯

**Pull request merged**